File: src/node/halchemy/src/lib/metadata.ts
import {doSettingsIncludeStatusCode} from "./status_codes";

export class HalchemyMetadata {
    public request: any
    public response: any
    public error: any

    constructor(request: any, response: any, error: any) {
        this.request = request
        this.response = response
        this.error = error
    }

    public raiseForStatusCodes(settings: string = '>399') {
        const statusCode = this.response.statusCode
        const shouldThrow = doSettingsIncludeStatusCode(settings, statusCode)
        if (shouldThrow) {
            throw this
        }
    }
}

File: src/ruby/halchemy/lib/halchemy/metadata.rb
# frozen_string_literal: true

module Halchemy
  # Allows a Resource to act first as a Hash, by keeping all HTTP related info in resource._halchemy
  class Metadata
    attr_reader :request, :response, :error

    def initialize(request, response, error)
      @request = request
      @response = response
      @error = error
    end

    def raise_for_status_codes(settings = ">399")
      return unless settings_include_status_code?(settings, @response.status_code)

      raise HttpError, "Status code #{@response.status_code} matches \"#{settings}\""
    end
  end
end
